commons/src/main/java/io/aiven/kafka/tieredstorage/commons/UniversalRemoteStorageManager.java

```diff
@@ -16,54 +16,223 @@
+            );
+            aesEncryptionProvider = new AesEncryptionProvider();
+        }
+        chunkManager = new ChunkManager(objectStorageFactory, objectKey, aesEncryptionProvider);
+        chunkSize = config.chunkSize();
+        compression = config.compressionEnabled();
+
+        mapper = getObjectMapper();
+    }
+
+    private ObjectMapper getObjectMapper() {
+        final ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.registerModule(new Jdk8Module());
+        if (encryption) {
+            final SimpleModule simpleModule = new SimpleModule();
+            simpleModule.addSerializer(SecretKey.class, new DataKeySerializer(rsaEncryptionProvider::encryptDataKey));
+            simpleModule.addDeserializer(SecretKey.class, new DataKeyDeserializer(
+                b -> new SecretKeySpec(rsaEncryptionProvider.decryptDataKey(b), "AES")));
+            objectMapper.registerModule(simpleModule);
+        }
+        return objectMapper;
     }
 
     @Override
     public void copyLogSegmentData(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
                                    final LogSegmentData logSegmentData) throws RemoteStorageException {
-        throw new UnsupportedOperationException();
+        Objects.requireNonNull(remoteLogSegmentMetadata, "remoteLogSegmentId must not be null");
+        Objects.requireNonNull(logSegmentData, "logSegmentData must not be null");
+        try {
+            TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(
+                Files.newInputStream(logSegmentData.logSegment()), chunkSize);
+            SegmentEncryptionMetadataV1 encryptionMetadata = null;
+            if (compression) {
+                transformEnum = new CompressionChunkEnumeration(transformEnum);
+            }
+            if (encryption) {
+                final DataKeyAndAAD dataKeyAndAAD = aesEncryptionProvider.createDataKeyAndAAD();
+                transformEnum = new EncryptionChunkEnumeration(
+                    transformEnum,
+                    () -> aesEncryptionProvider.encryptionCipher(dataKeyAndAAD));
+                encryptionMetadata = new SegmentEncryptionMetadataV1(dataKeyAndAAD.dataKey, dataKeyAndAAD.aad);
+            }
+            final var transformFinisher =
+                new TransformFinisher(transformEnum, remoteLogSegmentMetadata.segmentSizeInBytes());
+            try (final var sis = new SequenceInputStream(transformFinisher)) {
+                final String fileKey =
+                    objectKey.key(remoteLogSegmentMetadata, ObjectKey.Suffix.LOG);
+                objectStorageFactory.fileUploader().upload(sis, fileKey);
+            }
+
+
+            final ChunkIndex chunkIndex = transformFinisher.chunkIndex();
+            final SegmentManifest segmentManifest =
+                new SegmentManifestV1(chunkIndex, compression, encryptionMetadata);
+            uploadManifest(remoteLogSegmentMetadata, segmentManifest);
```

Comment:
minor: could we move this into its own method? maybe `uploadLogSegmentAndManifest` or similar?

Reply:
I was thinking that we will make the refactoring that we were planning with you(the pipelines) and refactor this at that time. Moving it to a separate method before does not bring too much value IMO but keeping it here can work as a reminder :)